### PR TITLE
Thales Royale private rooms as proxy and room management

### PR DESCRIPTION
**Pull request by grujawork**
[TIP-14](https://github.com/thales-markets/thales-improvement-proposals/blob/main/TIPs/TIP-14.md)

### CODE REVIEW

File: contracts/ThalesRoyale/ThalesRoyalePrivateRoom.sol [hunk 1]
```diff
@@ -199,46 +278,56 @@ contract ThalesRoyalePrivateRoom is Owned, Pausable {
...
+            "Round positioning finished"
+        );
 
         // this block is when sender change positions in a round - first reduce
-        if(positionInARoundPerRoom[_roomNumber][msg.sender][currentRoundInRoom[_roomNumber]] == 1){
-            positionsPerRoundPerRoom[_roomNumber][currentRoundInRoom[_roomNumber]][1]--;
-        }else if (positionInARoundPerRoom[_roomNumber][msg.sender][currentRoundInRoom[_roomNumber]] == 2) {
-            positionsPerRoundPerRoom[_roomNumber][currentRoundInRoom[_roomNumber]][2]--;
+        if (positionInARoundPerRoom[_roomNumber][msg.sender][currentRoundInRoom[_roomNumber]] == DOWN) {
+            positionsPerRoundPerRoom[_roomNumber][currentRoundInRoom[_roomNumber]][DOWN]--;
```
Comment (kirilantevski) [inline contracts/ThalesRoyale/ThalesRoyalePrivateRoom.sol:306]: Please use `.add()` and `.sub()` for math operation. 
The math library is imported, and protects from overflows.

File: contracts/ThalesRoyale/ThalesRoyalePrivateRoom.sol [hunk 2]
```diff
@@ -247,324 +336,395 @@ contract ThalesRoyalePrivateRoom is Owned, Pausable {
...
-            rewardToken.transfer(msg.sender, rewardPerPlayerPerRoom[_roomNumber]);
-
-            // emit event
-            emit RewardClaimed(_roomNumber, msg.sender, rewardPerPlayerPerRoom[_roomNumber]);
-        }
+        require(!rewardCollectedPerRoom[_roomNumber][msg.sender], "Player already collected reward");
+
+        // set collected -> true
+        rewardCollectedPerRoom[_roomNumber][msg.sender] = true;
+        unclaimedRewardPerRoom[_roomNumber] = unclaimedRewardPerRoom[_roomNumber].sub(rewardPerWinnerPerRoom[_roomNumber]);
```
Comment (kirilantevski) [inline contracts/ThalesRoyale/ThalesRoyalePrivateRoom.sol:426]: the comment from above [ use add() or sub() ]  is implemented here. 
I think is good to be everywhere

**Review by kirilantevski — CHANGES_REQUESTED** [review]

File: contracts/ThalesRoyale/ThalesRoyalePrivateRoom.sol [hunk 3]
```diff
@@ -130,67 +160,116 @@ contract ThalesRoyalePrivateRoom is Owned, Pausable {
...
         _buyIn(msg.sender, _roomNumber, buyInPerPlayerRerRoom[_roomNumber]);
 
-        if(!playerStartedSignUp[_roomNumber]){
-            playerStartedSignUp[_roomNumber] = true;
-        }
-
         emit SignedUpInARoom(msg.sender, _roomNumber);
     }
 
     function startRoyaleInRoom(uint _roomNumber) external onlyRoomParticipantes(_roomNumber) {
```
[the remark this answers is not in the document]
Comment (grujawork) [reply contracts/ThalesRoyale/ThalesRoyalePrivateRoom.sol:266]: done.

File: contracts/ThalesRoyale/ThalesRoyalePrivateRoom.sol (hunk 1 above)
Comment (grujawork) [reply contracts/ThalesRoyale/ThalesRoyalePrivateRoom.sol]: done. 👍

File: contracts/ThalesRoyale/ThalesRoyalePrivateRoom.sol [hunk 4]
```diff
@@ -49,75 +66,88 @@ contract ThalesRoyalePrivateRoom is Owned, Pausable {
     mapping(uint => uint) public roundTargetPriceInRoom;
```
[the remark this answers is not in the document]
Comment (grujawork) [reply contracts/ThalesRoyale/ThalesRoyalePrivateRoom.sol:66]: yep, had that in my mind but could not make proxy work with struct 🤷‍♂️

**Review by kirilantevski — APPROVED** [review]

**Pull request merged**